validation/validate.ts

```diff
@@ -0,0 +1,329 @@
+}
+
+/**
+ * Get all nodes that are connected to nodeId by label references.
+ * The nodes returned by this function are the roots of any
+ * DOM subtrees that are relevant to the computation of nodeId's
+ * accName.
+ * @param {CDPSession} client
+ * @param {DOM.backendNodeId} nodeId
+ */
+async function getAllNodesInComputation(
+  client: CDPSession,
+  nodeId: Protocol.DOM.BackendNodeId
+): Promise<Protocol.DOM.BackendNodeId[]> {
+  const stack: Protocol.DOM.BackendNodeId[] = [];
+  const relatedNodeIds: Protocol.DOM.BackendNodeId[] = [];
+  stack.push(nodeId);
+  // Iterative DFS traverses nodes connected by label references
+  while (stack && stack.length > 0) {
+    const currentNodeId = stack.pop();
+    relatedNodeIds.push(currentNodeId!);
+    const axTree = (await client.send('Accessibility.getPartialAXTree', {
+      backendNodeId: currentNodeId,
+    })) as Protocol.Accessibility.GetPartialAXTreeResponse;
+    // Nodes in axTree.nodes are in order of decreasing depth in the DOM.
+    // We traverse all nodes descendant of currentNodeId, checking whether
+    // each one contains a label reference.
+    for (const axNode of axTree.nodes) {
+      let nextRelatedNodes: Protocol.Accessibility.AXRelatedNode[] = [];
+      const sources: Protocol.Accessibility.AXValueSource[] | null =
+        axNode.name?.sources ?? null;
+      // Check if axNode is aria-labelledby any other nodes
+      if (
+        sources &&
+        sources.length > 0 &&
+        sources[0].attributeValue?.relatedNodes
+      ) {
+        nextRelatedNodes = sources[0].attributeValue.relatedNodes;
+      }
+      // Check if axNode is native <label>led by any other nodes
+      else if (
+        sources &&
+        sources.length > 2 &&
+        sources[2].nativeSourceValue?.relatedNodes
+      ) {
+        nextRelatedNodes = sources[2].nativeSourceValue.relatedNodes;
+      }
+
+      for (const relatedNode of nextRelatedNodes) {
+        if (!relatedNodeIds.includes(relatedNode.backendDOMNodeId)) {
+          stack.push(relatedNode.backendDOMNodeId);
+        }
+      }
+      // Stop iterating when we reach the current node :
+      // we have checked all descendants.
+      if (axNode.backendDOMNodeId === currentNodeId) {
+        break;
+      }
+    }
+  }
+  return relatedNodeIds;
+}
+
+/**
+ * Removes any node in relatedNodeIds that is descendant
+ * of another node in relatedNodeIds. Such a node would be included
+ * twice in the minimised HTML because outerHTML includes
+ * all descendants.
+ * @param {CDPSession} client
+ * @param {DOM.backendNodeId} relatedNodeIds
+ */
+async function removeRedundantNodes(
+  client: CDPSession,
+  relatedNodeIds: Protocol.DOM.BackendNodeId[]
+): Promise<Protocol.DOM.BackendNodeId[]> {
+  const redundantNodeIds: Protocol.DOM.BackendNodeId[] = [];
+  for (const currentBackendNodeId of relatedNodeIds) {
+    const axTree = (await client.send('Accessibility.getPartialAXTree', {
+      backendNodeId: currentBackendNodeId,
```

Comment:
Rather than using the a11y tree to check the heirarchy, I'd use `Node.contains()`. It saves re-calculating the a11y tree for each element, and you need to convert to `HTMLElement` anyway to rebuild the DOM

Reply:
I don't think we can get `Node`s or `HTMLElement`s from the devtools protocol. I never actually work with `HTMLElement`s in this code, but rather use devtools `DOM.getOuterHTML` to get the markup.

However, [`DOM.querySelectorAll`](https://chromedevtools.github.io/devtools-protocol/tot/DOM/#method-querySelectorAll) might be a nicer way to do this as it returns an array of `NodeId`s, might be easier (and more intuitive) than calculating the a11y tree each time?